rllib/evaluation/env_runner_v2.py

```diff
@@ -379,68 +379,75 @@ def run(self) -> Iterator[SampleBatchType]:
             Object containing state, action, reward, terminal condition,
             and other fields as dictated by `policy`.
         """
-        # Before the very first poll (this will reset all vector sub-environments):
-        # Create all upcoming episodes and call `on_episode_created` callbacks for
-        # all sub-environments (upcoming episodes).
-        for env_id, sub_env in self._base_env.get_sub_environments(
-            as_dict=True
-        ).items():
-            self.create_episode(env_id)
-
         while True:
-            self._perf_stats.incr("iters", 1)
-
-            t0 = time.time()
-            # Get observations from all ready agents.
-            # types: MultiEnvDict, MultiEnvDict, MultiEnvDict, MultiEnvDict, ...
-            (
-                unfiltered_obs,
-                rewards,
-                dones,
-                infos,
-                off_policy_actions,
-            ) = self._base_env.poll()
-            env_poll_time = time.time() - t0
-
-            # Process observations and prepare for policy evaluation.
-            t1 = time.time()
-            # types: Set[EnvID], Dict[PolicyID, List[AgentConnectorDataType]],
-            #       List[Union[RolloutMetrics, SampleBatchType]]
-            to_eval, outputs = self._process_observations(
-                unfiltered_obs=unfiltered_obs,
-                rewards=rewards,
-                dones=dones,
-                infos=infos,
-            )
-            self._perf_stats.incr("raw_obs_processing_time", time.time() - t1)
-
+            outputs = self.step()
             for o in outputs:
                 yield o
 
-            # Do batched policy eval (accross vectorized envs).
-            t2 = time.time()
-            # types: Dict[PolicyID, Tuple[TensorStructType, StateBatch, dict]]
-            eval_results = self._do_policy_eval(to_eval=to_eval)
-            self._perf_stats.incr("inference_time", time.time() - t2)
-
-            # Process results and update episode state.
-            t3 = time.time()
-            actions_to_send: Dict[
-                EnvID, Dict[AgentID, EnvActionType]
-            ] = self._process_policy_eval_results(
-                to_eval=to_eval,
-                eval_results=eval_results,
-                off_policy_actions=off_policy_actions,
-            )
-            self._perf_stats.incr("action_processing_time", time.time() - t3)
+    def step(self) -> List[SampleBatchType]:
```

Comment:
Moved this to simplify writing tests that require us only to step through an env with a policy in the future (so that we can get Algorithm out of the picture for such tests)

Reply:
nice!!